rfcs/0004-fulltext-search.md

```diff
@@ -0,0 +1,236 @@
+the subgraph's definition. The subgraph developer will create a new virtual 
+field of type String that is marked with the special `@index` directive. 
+The directive will have two required parameters for specifying the index
+`type` and `params`. Params fields are specific to the index type; in the 
+case of fulltext search there will be two params to specify: `language` 
+and `fields`. Verification of the index definition will ensure that all 
+fields referenced are valid String type fields. To be clear, the proposed 
+interface will add the fulltext search index definitions to the subgraph 
+deployment hash, so adding or updating a fulltext search index will require 
+an update to the subgraph.  With subgraph composition it will be possible 
+to easily create new subgraphs that add specific fulltext search indexes 
+to an existing subgraph. 
+
+Example fulltext search field definition:
+```graphql
+type ExampleItem @entity {
+    id: ID!,
+    name: String!,
+    description: String!,
+    specs: String!,
+    address: Bytes!,
+    marked: Boolean!,
+    itemSearch: String! @index(
+        type: "fulltext", 
+        params: {
+            language:"english", 
+            fields: [
+                {name: "name", weight: 5}, 
+                {name: "description", weight: 3}, 
+                {name: "specs", weight: 1},                
+            ],            
+        }
+    )
+}
+```  
+
+### GraphQL Query interface
+
+For each fulltext index three search algorithms will be made available to 
+the end user: a simple text query, a standard ranking, and a cover 
+density ranking algorithm: 
+  - `query`: search for the string in the indexed document. Several 
+    operators are available: and, or, and proximity (`&`, `|`, `<->`.) 
+    The proximity operator allows one to specify max distance: 
+    3 words apart → `<3>`. (no filter suffix)
+  - `standard ranking`: ranking based on the number of matching lexemes. 
+    (filter suffix = `_rank`)
+  - `cover density ranking`: Cover density is similar to the standard 
+    fulltext search ranking except that the proximity of matching lexemes 
+    to each other is taken into consideration. This function requires 
+    lexeme positional information to perform its calculation, so it ignores 
+    any "stripped" lexemes in the index.
+    (filter suffix = `_proximity_rank`)
+    
+Example Usage in a query (references the example Entity definition from above section): 
+```graphql
+query {
+    exampleItems(first: 10, where: {itemSearch_proximity_rank: "sarcastic&tea"}) {
+        address
+        name 
+        description
+    }
+}
+```
+
+### Tools and Design
+
+Fulltext search query system implementations often involve specific systems 
+for storing and querying the text documents; however, in an effort to reduce 
+system complexity and feature implementation time I propose starting with the 
+fulltext search features built in to PostgreSQL.
```

Comment:
When we build query verification into the protocol, we may not be able to use Postgres for storing entity data anymore. If we build Postgres-specific fulltext search features into the protocol _now_, we may likely have to break the protocol later. @yanivtal @Zerim Thoughts?

Reply:
Do we have any choice here? Are there full-text search features that are common and deterministic to all SQL databases.

For now, I'm OK w/ us having SQL or even Postgres specific full-text search functionality in the protocol w/ the understanding that there's a lot that may have to change when we eventually move to index chains and V2.